apisix/plugins/hmac-auth.lua

```diff
@@ -0,0 +1,303 @@
+    priority = 2530,
+    type = 'auth',
+    name = plugin_name,
+    schema = schema,
+}
+
+local hmac_funcs = {
+    ["hmac-sha1"] = function(secret_key, message)
+        return hmac_sha1(secret_key, message)
+    end,
+    ["hmac-sha256"] = function(secret_key, message)
+        return hmac:new(secret_key, hmac.ALGOS.SHA256):final(message)
+    end,
+    ["hmac-sha512"] = function(secret_key, message)
+        return hmac:new(secret_key, hmac.ALGOS.SHA512):final(message)
+    end,
+}
+
+
+local function try_attr(t, ...)
+    local tbl = t
+    local count = select('#', ...)
+    for i = 1, count do
+        local attr = select(i, ...)
+        tbl = tbl[attr]
+        if type(tbl) ~= "table" then
+            return false
+        end
+    end
+
+    return true
+end
+
+
+local create_consumer_cache
+do
+    local consumer_ids = {}
+
+    function create_consumer_cache(consumers)
+        core.table.clear(consumer_ids)
+
+        for _, consumer in ipairs(consumers.nodes) do
+            core.log.info("consumer node: ", core.json.delay_encode(consumer))
+            consumer_ids[consumer.auth_conf.access_key] = consumer
+        end
+
+        return consumer_ids
+    end
+
+end -- do
+
+
+function _M.check_schema(conf)
+    core.log.info("input conf: ", core.json.delay_encode(conf))
+
+    return core.schema.check(schema, conf)
+end
+
+
+local function get_consumer(access_key)
+    if not access_key then
+        return nil, {message = "missing access key"}
+    end
+
+    local consumer_conf = consumer.plugin(plugin_name)
+    if not consumer_conf then
+        return nil, {message = "Missing related consumer"}
+    end
+
+    local consumers = core.lrucache.plugin(plugin_name, "consumers_key",
+            consumer_conf.conf_version,
+            create_consumer_cache, consumer_conf)
+
+    local consumer = consumers[access_key]
+    if not consumer then
+        return nil, {message = "Invalid access key"}
+    end
+    core.log.info("consumer: ", core.json.delay_encode(consumer))
+
+    return consumer
+end
+
+
+local function generate_signature(ctx, secret_key, params)
+    local canonical_uri = ctx.var.uri
+    local canonical_query_string = ""
+    local request_method = ngx_req.get_method()
+    local args = ngx_req.get_uri_args()
+
+    if canonical_uri == "" then
+        canonical_uri = "/"
+    end
+
+    if type(args) == "table" then
+        local keys = {}
+        local query_tab = {}
+
+        for k, v in pairs(args) do
+            core.table.insert(keys, k)
+        end
+        core.table.sort(keys)
+
+        for _, key in pairs(keys) do
+            local param = args[key]
+            if type(param) == "table" then
+                for _, vval in pairs(param) do
+                    core.table.insert(query_tab, escape_uri(key) .. "=" .. escape_uri(vval))
+                end
+            else
+                core.table.insert(query_tab, escape_uri(key) .. "=" .. escape_uri(param))
+            end
```

Comment:
The encoding process of these two branches, which is the "key" method and which is the "key=value" method? I don't quite understand here.

Reply:
> The encoding process of these two branches, which is the "key" method and which is the "key=value" method? I don't quite understand here.

please take a look at `ngx_req.get_uri_args` API first.

Reply:
Thanks, I understand.